File: 0904_fruitIntoBaskets/main.py
from typing import List
from collections import defaultdict

class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        count = defaultdict(int)
        l, res = 0, 0
        for r in range(len(fruits)):
            count[fruits[r]] += 1

            while len(count) > 2:
                count[fruits[l]] -= 1
                if count[fruits[l]] == 0:
                    count.pop(fruits[l])
                l += 1
            
            res = max(res, r - l + 1)
        return res

File: 0904_fruitIntoBaskets/main.cpp
#include <vector>
#include <unordered_map>
#include <algorithm>

class Solution {
public:
    int totalFruit(std::vector<int>& fruits) {
        std::unordered_map<int, int> count;
        int l = 0;
        int res = 0;
        for(int r = 0; r < fruits.size(); ++r) {
            ++count[fruits[r]];
            while(count.size() > 2) {
                --count[fruits[l]];
                if(count[fruits[l]] == 0)
                    count.erase(fruits[l]);
                ++l;
            }
            res = std::max(res, r - l + 1);
        }
        return res;
    }
};